packages/core/contracts/optimistic-oracle-v2/implementation/WhitelistOptimisticOracleV2.sol

```diff
@@ -0,0 +1,275 @@
+     * @param timestamp timestamp to identify the existing request.
+     * @param ancillaryData ancillary data of the price being requested.
+     */
+    function ownerSetEventBased(
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData
+    ) external nonReentrant() onlyOwner() {
+        require(
+            _getState(requester, identifier, timestamp, ancillaryData) == State.Requested,
+            "setEventBased: Requested"
+        );
+        Request storage request = _getRequest(requester, identifier, timestamp, ancillaryData);
+        request.requestSettings.eventBased = true;
+        request.requestSettings.refundOnDispute = true;
+    }
+
+    /**
+     * @notice Sets which callbacks should be enabled for the request.
+     * @param requester sender of the initial price request.
+     * @param identifier price identifier to identify the existing request.
+     * @param timestamp timestamp to identify the existing request.
+     * @param ancillaryData ancillary data of the price being requested.
+     * @param callbackOnPriceProposed whether to enable the callback onPriceProposed.
+     * @param callbackOnPriceDisputed whether to enable the callback onPriceDisputed.
+     * @param callbackOnPriceSettled whether to enable the callback onPriceSettled.
+     */
+    function ownerSetCallbacks(
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData,
+        bool callbackOnPriceProposed,
+        bool callbackOnPriceDisputed,
+        bool callbackOnPriceSettled
+    ) external nonReentrant() onlyOwner() {
+        require(
+            _getState(requester, identifier, timestamp, ancillaryData) == State.Requested,
+            "setCallbacks: Requested"
+        );
+        Request storage request = _getRequest(requester, identifier, timestamp, ancillaryData);
+        request.requestSettings.callbackOnPriceProposed = callbackOnPriceProposed;
+        request.requestSettings.callbackOnPriceDisputed = callbackOnPriceDisputed;
+        request.requestSettings.callbackOnPriceSettled = callbackOnPriceSettled;
+    }
+
+    /**
+     * @notice Sets the proposer whitelist for a request.
+     * @param requester sender of the initial price request.
+     * @param identifier price identifier to identify the existing request.
+     * @param timestamp timestamp to identify the existing request.
+     * @param ancillaryData ancillary data of the price being requested.
+     * @param whitelist address of the whitelist to set.
+     */
+    function ownerSetProposerWhitelist(
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData,
+        address whitelist
+    ) external nonReentrant() onlyOwner() {
+        customProposerWhitelists[_getId(requester, identifier, timestamp, ancillaryData)] = AddressWhitelistInterface(
```

Comment:
It might make sense to also check if the given request state is `State.Requested`. This could help the caller be better assured that the applied whitelist would in fact be effective. Otherwise they could be under impression that the whitelist was set, but in fact the proposal has been already made against the default whitelist.